sound library banks: add an underscore between name and bank, like in strudel.

telephon opened the pull request:
As discussed here: #231
It may break code, but since the feature was unknown, and easy to fix, it should be good.

I've added a convenience method, you can now write: 
```supercollider
~dirt.loadSoundFilesToBank(bankName: "fufu");
```

Then in tidal:
```haskell
s "bd sd rd" # bank "fufu"
```

or synonymously:

```haskell
s "fufu_bd fufu_sd fufu_rd" 
```

telephon committed:
add an underscore between name and bank, like in strudel

telephon committed:
sound library: loadSoundFilesToBank

This is for convenience and encapsulation of the naming standard.

telephon committed:
sound library: convenience method for loading banks

yaxu commented:
@geikha 's [tidal-drum-machines project](https://github.com/geikha/tidal-drum-machines/) uses `-` as the delimiter in the filesystem.

Strudel [switches this to `_` in the mapping](https://github.com/tidalcycles/strudel/blob/9b4574c7b0dac3627e59207adfc560570b7a7885/website/public/tidal-drum-machines.json#L4), probably because `-` isn't a  valid identifier char in the mininotation

I'm not sure how best to fix this.. Could do it in the filenames in the repo but would have to fix strudel at the same time.  Could also just duplicate the repo.. Otherwise SuperDirt could read the same json file format.

(It also [defines some bank shorthands](https://github.com/tidalcycles/strudel/blob/9b4574c7b0dac3627e59207adfc560570b7a7885/website/public/tidal-drum-machines-alias.json) so e.g. `RolandTR909` can be specified just as `tr909`)

yaxu commented on classes/DirtEvent.sc:
```diff
@@ -37,7 +37,9 @@ DirtEvent {
 		var s, n;
 		#s, n = ~sound.asString.split($:);
-		if(~bank.notNil) { s = ~bank ++ s };
+		if(~bank.notNil) {
```
I think splitName isn't called if `s` is defined, and tidal sends `s` rather than `sound`, so this isn't reached.

telephon replied:
ah that's true. I'll have to do it a little difefrently then.

telephon committed:
dirt event: make sure ~s also gets a sample bank

telephon commented:
Can you test this one?

telephon commented:
> Otherwise SuperDirt could read the same json file format.

Yes, sure, it could. But unless there is a really clear standard, this just adds complexity ...

telephon commented:
I can add a delimiter argument to `loadSoundFilesToBank`, or just ask the user to append the delimiter to the name:

```
~dirt.loadSoundFilesToBank(bankName: "fufu_");
```

yaxu commented:
Yes that works for me thanks!

I think it is probably easier to just rearrange the samples rather than make a level of indirection for now. It seems symlinks work so we could do it that way.

telephon commented:
> I think it is probably easier to just rearrange the samples rather than make a level of indirection for now. It seems symlinks work so we could do it that way.

How would you have tio change them?

yaxu commented:
Mainly just change `-` for  `_`.  https://github.com/geikha/tidal-drum-machines/tree/main/machines/RolandTR909

The folder names are also quite long, e.g. `tr909` would be better in this case.

If  a `namingFunction` was passed through to `loadSoundFiles`, then that would allow a lot of flexibility, something like this:

```
~dirt.loadSoundFilesToBank('./machines/RolandTR909', bankName: "tr909", namingFunction: {arg x; x.split('-').pop()});
```

Then we could just add a `.scd` file to repos like `tidal-drum-machines` that can be included in startups to get all the samples. I'm not sure how to express a sample path that would work inside the .scd file though.. with `nowExecutingPath`?

telephon commented:
Also, this has slipped under the radar, I'm sorry. What si the status @yaxu, we should still do it as you suggested?

geikha commented:
Woah just saw this too! I agree the names are long, I had done a survey back then and people majorly voted for full names, that's why I had them like that. To me the best case scenario would be developing some sort of info file that we can add to folders for aliases and other stuff

telephon committed:
DirtSoundLibrary: pass on naming function when reading bank

telephon commented:
> To me the best case scenario would be developing some sort of info file that we can add to folders for aliases and other stuff

That would be easy once we know what we want.

telephon committed:
DirtSoundLibrary: fix typo

telephon commented:
> If a `namingFunction` was passed through to `loadSoundFiles`, then that would allow a lot of flexibility, something like this:
> 
> ```
> ~dirt.loadSoundFilesToBank('./machines/RolandTR909', bankName: "tr909", namingFunction: {arg x; x.split('-').pop()});
> ```

I've made a change so the namingFunction is now passed through. What is not so practical right now is that if you pass it in you have to already know the bank name. We could pass the name to the function to curry out another function to pass on.

One question, when you wrote: `{ arg x; x.split('-').pop() }` this would require dashes in all names. I suppose you wanted something that always works, right?


> Then we could just add a `.scd` file to repos like `tidal-drum-machines` that can be included in startups to get all the samples. I'm not sure how to express a sample path that would work inside the .scd file though.. with `nowExecutingPath`?

Yes, almost. We can should `resolveRelative` to resolve the path relative to the file we are in.

geikha commented:
The original drum machine repo uses `{|x| x.basename.replace("-","")}` as its `namingFunction`. Btw, since I'm clarifying some of the original desicions, the reason I added a `-` in between all drum machines was to make it easy to rename banks or elements (I can imagine someone downloading the repo and changing all `sd` to `sn`).

---

I'd imagine the info file having the following (pseudocode ahead, i'll use the TR-808 as an example):
```
name: "rolandtr808"
type: bank // could be either "bank" or "folder"
aliases: ["tr808", "808"]
ignore: false // allows you to tell SD to ignore the bank/folder without deleting it
```
For folders maybe we could have explicit orders. Also wondering how such a system could handle the `sd` vs `sn` scenario I mentioned above without manually adding an alias to all the `sd` folders.

bsssssss commented on classes/DirtEvent.sc:
```diff
@@ -34,12 +33,19 @@ DirtEvent {
+			sound = format("%_%", ~bank, sound)
+		};
+		~s = sound.asSymbol;
+		~n = if(note.notNil) { note.asFloat } { 0.0 };
```
While testing this i noticed sample selection was broken (seems like it always load `sample:0.0` regardless the value of `~n`).
I think `~n = if(note.notNil) { note.asFloat } { ~n };` fixes it ?

telephon replied:
oh yes, of course (a typo). Thanks!

telephon committed:
dirt-event: fix typo, keep ~n

az4g commented:
Any chance for this PR to be merged someday? I added it to my local install, it works like a charm and could be useful to other users.

telephon commented:
Ah it works? So this isn't a problem?: https://github.com/musikinformatik/SuperDirt/pull/312/files#r2021736182

telephon commented:
OK, I understood, yes it is fixed.

az4g commented:
Thanks!
